### ISSUE TITLE
display  progress 

### ISSUE THREAD

**Issue by zszq** [graphology/graphology-layout-forceatlas2#24]
I want to know the current calculation progress, which is used to display the progress bar. Is there any good way?

**Comment by Yomguithereal** [comment]
Hello @zszq. There is no way to know how many iterations of the algorithms are required to finish layout. Sometimes it might never finish nor converge, like many existing iterative algorithms such as k-means. Usually people let their users start/stop the layout to let them make this choice by observing the algorithm while it runs.

**Comment by zszq** [comment]
> Hello @zszq. There is no way to know how many iterations of the algorithms are required to finish layout. Sometimes it might never finish nor converge, like many existing iterative algorithms such as k-means. Usually people let their users start/stop the layout to let them make this choice by observing the algorithm while it runs.
> 
> You can also arbitrarily run a fixed number of iterations of the layout algorithm on your graph but you will not have any proof that it converged or that it can converge.



> Hello @zszq. There is no way to know how many iterations of the algorithms are required to finish layout. Sometimes it might never finish nor converge, like many existing iterative algorithms such as k-means. Usually people let their users start/stop the layout to let them make this choice by observing the algorithm while it runs.
> 
> You can also arbitrarily run a fixed number of iterations of the layout algorithm on your graph but you will not have any proof that it converged or that it can converge.

Thank you for your reply. When I have 5000 points and 5000 lines, I used the `#.inferSettings` method. The calculation takes more than 5 minutes. Is this normal? I only need the final layout position and display it on the canvas. Is there any better way to speed up the calculation?

**Comment by Yomguithereal** [comment]
> The calculation takes more than 5 minutes. Is this normal?

It mostly depends on your machine but yes it could be. Can you check that `#.inferSettings` is returning `barnesHutOptimize` set to `true`? If this is not the case, set this parameter to `true`. It will optimize iterations from `O(n^2)` to `O(n logn)`.

> I only need the final layout position and display it on the canvas. Is there any better way to speed up the calculation?

Maybe 5000 iterations is a bit much. You can surely try lmuch ess and see if it works for you. A force layout is basically a solver for the n-body problem and requires to apply repulsion on each pair of node. This can be optimized through Barnes-Hut calculation as suggested above, but it is hard to be more efficient that this. Note also that those algorithms where mostly designed having scale-free networks in mind rather than mathematical ones. Typically it does not work very well on lattices, grids, or binomial graph etc. but I don't really know which kind of graphs you are handling so I cannot say whether this particular layout is a good idea or not in your case.

**Comment by zszq** [comment]
> > 计算时间超过 5 分钟。这是正常的吗？
> 
> 这主要取决于您的机器，但它可能是。你能检查一下`#.inferSettings`返回`barnesHutOptimize`设置为`true`吗？如果不是这种情况，请将此参数设置为`true`。它将优化从`O(n^2)`到 的迭代`O(n logn)`。
> 
> > 我只需要最终的布局位置并将其显示在画布上。有没有更好的方法来加速计算？
> 
> 也许 5000 次迭代有点多。你当然可以少尝试，看看它是否适合你。力布局基本上是 n 体问题的求解器，需要在每对节点上施加排斥力。这可以通过如上所述的 Barnes-Hut 计算进行优化，但很难比这更有效。另请注意，那些设计时主要考虑的是无标度网络而不是数学算法的算法。通常它在格子、网格或二项式图等上效果不佳，但我真的不知道您正在处理哪种图形，所以我不能说这种特定的布局在您的情况下是否是一个好主意。



> > The calculation takes more than 5 minutes. Is this normal?
> 
> It mostly depends on your machine but yes it could be. Can you check that `#.inferSettings` is returning `barnesHutOptimize` set to `true`? If this is not the case, set this parameter to `true`. It will optimize iterations from `O(n^2)` to `O(n logn)`.
> 
> > I only need the final layout position and display it on the canvas. Is there any better way to speed up the calculation?
> 
> Maybe 5000 iterations is a bit much. You can surely try much less and see if it works for you. A force layout is basically a solver for the n-body problem and requires to apply repulsion on each pair of node. This can be optimized through Barnes-Hut calculation as suggested above, but it is hard to be more efficient that this. Note also that those algorithms where mostly designed having scale-free networks in mind rather than mathematical ones. Typically it does not work very well on lattices, grids, or binomial graph etc. but I don't really know which kind of graphs you are handling so I cannot say whether this particular layout is a good idea or not in your case.

thanks for your advice! Does graphology include a hierarchical layout algorithm for network graphs up and down and left and right? I don’t seem to find this kind of library. I’m not familiar with this aspect. Is there any good library recommendation?

**Comment by Yomguithereal** [comment]
> Does graphology include a hierarchical layout algorithm for network graphs up and down and left and right?

Not yet unfortunately. I don't know your use case but people tend to use this with tree (acyclical graphs) rather than graphs, otherwise it can be a bit tricky to make decisions about your data's hierarchy. People usually use d3 for this kind of trees though.

**Comment by zszq** [comment]
> > Does graphology include a hierarchical layout algorithm for network graphs up and down and left and right?
> 
> Not yet unfortunately. I don't know your use case but people tend to use this with tree (acyclical graphs) rather than graphs, otherwise it can be a bit tricky to make decisions about your data's hierarchy. People usually use d3 for this kind of trees though.

I want to add many points or lines at once and then trigger the event only once instead of triggering multiple times? Why can't `addNodesFrom `be used?

**Comment by Yomguithereal** [comment]
The `addNodesFrom` does not exist anymore. Just loop over your source collection and add nodes one by one. I am not sure to understand which event you are speaking of? If you are speaking of `nodeAdded`, just debounce your action and you should be good to go.

**Comment by zszq** [comment]
> The `addNodesFrom` does not exist anymore. Just loop over your source collection and add nodes one by one. I am not sure to understand which event you are speaking of? If you are speaking of `nodeAdded`, just debounce your action and you should be good to go.

Listen to the `nodeAdded` event. When I add multiple nodes, I only want to execute the event once for the last time. Now it will execute multiple times.

**Comment by zszq** [comment]
> The `addNodesFrom` does not exist anymore. Just loop over your source collection and add nodes one by one. I am not sure to understand which event you are speaking of? If you are speaking of `nodeAdded`, just debounce your action and you should be good to go.

Listen to the `nodeAdded` event. When I add multiple nodes, I only want to execute the event once for the last time. Now it will execute multiple times.

**Comment by Yomguithereal** [comment]
Yes and this is how it should work. What you need is a debouncer. You have one in lodash for instance: https://lodash.com/docs/4.17.15#debounce

**Comment by zszq** [comment]
> Yes and this is how it should work. What you need is a debouncer. You have one in lodash for instance: https://lodash.com/docs/4.17.15#debounce

This is indeed a good idea, thank you
